File: TableRW.Epplus/Read/ExcelWorksheetEx.cs

using OfficeOpenXml;
using TableRW.Read.Epplus;
using TableRW.Utils.Ex;

namespace TableRW.Read;

public static class ExcelWorksheetEx {
    public static List<TEntity> ReadToList<TEntity>(
        this ExcelWorksheet sheet, int headerRow
    ) where TEntity : new() {
        if (sheet == null) { throw new ArgumentNullException(nameof(sheet)); }
        if (sheet.Dimension == null) { throw new ArgumentNullException("sheet.Dimension == null"); }

        if (CacheReadFn<TEntity>.FnUseHeader is var fn && fn != null) {
            return fn(sheet);
        }

        var header = GetHeader();
        if (header.Count == 0) {
            throw new InvalidOperationException("No data read: The number of column headers is 0");
        }

        var (iCol, m0) = header[0];
        var reader = new ExcelReader<TEntity>();
        reader.SetStart(headerRow + 1, iCol);
        reader.AddColumn(m0);

        foreach (var (i, m) in header.Skip(1)) {
            if (i - iCol > 1) {
                reader.AddSkipColumn(i - iCol - 1);
            }
            reader.AddColumn(m);
            iCol = i;
        }

        var readLmd = reader.Lambda();
        CacheReadFn<TEntity>.FnUseHeader = fn = readLmd.Compile();
        return fn(sheet);

        List<(int i, MemberInfo member)> GetHeader() {
            var t_entity = typeof(TEntity);
            var props = t_entity.GetProperties().Where(p => p.CanWrite)
                .Concat<MemberInfo>(t_entity.GetFields().Where(f => !f.IsInitOnly))
                .Where(m => m.HasAttribute<IgnoreReadAttribute>() == false)
                .ToDictionary(m => m.Name);

            return Range(1, sheet.Dimension.Columns + 1)
                .Select(i => (i, sheet.Cells[headerRow, i].Text))
                .Select((t) => (t.i, member: props.GetValueOr(t.Text, null!)))
                .Where(t => t.member != null)
                .ToList();
        }
    }

    public static List<TEntity> ReadToList<TEntity>(
        this ExcelWorksheet sheet,
        int cacheKey,
        Func<ExcelReader<TEntity>, Func<ExcelWorksheet, List<TEntity>>> buildRead
    ) where TEntity : new() {
        if (CacheReadFn<TEntity>.DicFn is var dic && !dic.TryGetValue(cacheKey, out var fn)) {
            dic[cacheKey] = fn = buildRead(new());
        }

        return fn(sheet);
    }

    public static List<TEntity> ReadToList<TEntity, TData>(
        this ExcelWorksheet sheet,
        int cacheKey,
        Func<ExcelReader<TEntity, TData>, Func<ExcelWorksheet, List<TEntity>>> buildRead
    ) where TEntity : new() {
        if (CacheReadFn<TEntity>.DicFn is var dic && !dic.TryGetValue(cacheKey, out var fn)) {
            dic[cacheKey] = fn = buildRead(new());
        }

        return fn(sheet);
    }

}

static class CacheReadFn<T> {
    internal static Func<ExcelWorksheet, List<T>>? FnUseHeader;

    internal static Dictionary<int, Func<ExcelWorksheet, List<T>>> DicFn = new();
}

File: TableRW.Epplus.Fs/Read/ExcelWorksheetEx.fs
namespace TableRW.Read.Epplus.Fs

open OfficeOpenXml
open TableRW.Read.Epplus
open TableRW
open TableRW.Utils.Ex
open TableRW.Read.Epplus.Fs
open System.Reflection
open System.Linq

module ExcelWorksheetEx =
    type ExcelWorksheet with
        member sheet.ReadToList<'TEntity> headerRow =
            let reader = new ExcelReader<'TEntity>()
            let t_entity = typeof<'TEntity>

            let props =
                t_entity.GetProperties().Where(fun p -> p.CanWrite).Cast<MemberInfo>()
                |> Seq.append (t_entity.GetFields().Where(fun f -> not f.IsInitOnly).Cast())
                |> Seq.filter (fun m -> not (m.HasAttribute<IgnoreReadAttribute>()))
                |> Seq.map (fun m -> m.Name, m) |> Map.ofSeq

            let header =
                { 1..sheet.Dimension.Columns }
                |> Seq.map (fun i -> i, sheet.Cells[headerRow, i].Text)
                |> Seq.map (fun (i, text) -> i, props.GetValueOr text null)
                |> Seq.filter (fun (i, m) -> m <> null)


            0
